### PR TITLE
core: several cleanups/fixes for WorkingDirectory= handling

### PR DESCRIPTION
**Pull request by YHNdnzj**
Fixes https://github.com/systemd/systemd/issues/33015

### CODE REVIEW

File: src/core/unit.c [hunk 1]
```diff
@@ -4238,16 +4238,14 @@ int unit_patch_contexts(Unit *u) {
                                         return -ENOMEM;
                         }
 
-                if (MANAGER_IS_USER(u->manager) &&
-                    !ec->working_directory) {
-
+                if (ec->working_directory_home || (MANAGER_IS_USER(u->manager) && !ec->working_directory)) {
                         r = get_home_dir(&ec->working_directory);
```
Comment (yuwata) [inline src/core/unit.c:4257]: Hm? If `WorkingDirectory=~` is specified in a system unit, then it always depends on the home directory of root(UID0)??

> WorkingDirectory=
           Sets the working directory for executed processes. If set to "~", the home directory of the user
           specified in User= is used.

File: src/core/exec-invoke.c [hunk 2]
```diff
@@ -3600,11 +3600,11 @@ static int acquire_home(const ExecContext *c, uid_t uid, const char** home, char
         if (!c->working_directory_home)
                 return 0;
 
-        r = get_home_dir(buf);
+        r = get_home_dir(ret_buf);
```
Comment (yuwata) [inline src/core/exec-invoke.c:3606]: At this stage, the uid is not applied yet, so get_home_dir() provides the home directory of the root, no?
Though, this is not your fault.

File: src/core/unit.c (hunk 1 above)
Comment (YHNdnzj) [reply src/core/unit.c:4257]: Hmm, right. I'll drop the change.

File: src/core/exec-invoke.c (hunk 2 above)
Comment (YHNdnzj) [reply src/core/exec-invoke.c:3606]: If `User=` is explicitly specified, the home returned by `get_fixed_user` will be used. I'll add a comment.

Comment (yuwata) [reply src/core/exec-invoke.c:3606]: Ah, indeed.

BTW, `get_fixed_user()` is (of course) called when DynamicUser= is disabled.
Maybe, we should refuse `DynamicUser=` with `WorkindDirectory=~` ?

File: src/core/exec-invoke.c [hunk 3]
```diff
@@ -3594,7 +3594,7 @@ static int acquire_home(const ExecContext *c, const char **home, char **ret_buf)
 
         /* If WorkingDirectory=~ is set, try to acquire a usable home directory. */
 
-        if (*home)
+        if (*home) /* Already acquired from get_fixed_user? */
```
Comment (yuwata) [inline src/core/exec-invoke.c]: We usually make function names suffixed with `()`.

File: src/core/exec-invoke.c (hunk 2 above)
Comment (YHNdnzj) [reply src/core/exec-invoke.c:3606]: Hmm, is `DynamicUser=` supposed to work with unprivileged users? It seems that the manager doesn't refuse the unit but sd-executor fails half-way, when trying to take a lock on `/etc/.pwd.lock`?

**Review by bluca — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Pull request merged**